adafruit_ble/advertising/__init__.py

```diff
@@ -0,0 +1,287 @@
+        self._advertisement = advertisement
+        self._adt = advertising_data_type
+        self.flags = None
+        if self._adt in self._advertisement.data_dict:
+            self.flags = self._advertisement.data_dict[self._adt][0]
+        elif self._advertisement.mutable:
+            self.flags = 0b110 # Default to General discovery and LE Only
+        else:
+            self.flags = 0
+
+    def __bytes__(self):
+        encoded = bytearray(1)
+        encoded[0] = self.flags
+        return encoded
+
+    def __str__(self):
+        parts = ["<AdvertisingFlags"]
+        for attr in dir(self.__class__):
+            attribute_instance = getattr(self.__class__, attr)
+            if issubclass(attribute_instance.__class__, AdvertisingFlag):
+                if getattr(self, attr):
+                    parts.append(attr)
+        parts.append(">")
+        return " ".join(parts)
+
+class String(AdvertisingDataField):
+    """UTF-8 encoded string in an Advertisement.
+
+       Not null terminated once encoded because length is always transmitted."""
+    def __init__(self, *, advertising_data_type):
+        self._adt = advertising_data_type
+
+    def __get__(self, obj, cls):
+        if self._adt not in obj.data_dict:
+            return None
+        return str(obj.data_dict[self._adt], "utf-8")
+
+    def __set__(self, obj, value):
+        obj.data_dict[self._adt] = value.encode("utf-8")
+
+class Struct(AdvertisingDataField):
+    """`struct` encoded data in an Advertisement."""
+    def __init__(self, struct_format, *, advertising_data_type):
+        self._format = struct_format
+        self._adt = advertising_data_type
+
+    def __get__(self, obj, cls):
+        if self._adt not in obj.data_dict:
+            return None
+        return struct.unpack(self._format, obj.data_dict[self._adt])[0]
+
+    def __set__(self, obj, value):
+        obj.data_dict[self._adt] = struct.pack(self._format, value)
+
+
+class LazyField(AdvertisingDataField):
+    """Non-data descriptor useful for lazily binding a complex object to an advertisement object."""
+    def __init__(self, cls, attribute_name, *, advertising_data_type, **kwargs):
+        self._cls = cls
+        self._attribute_name = attribute_name
+        self._adt = advertising_data_type
+        self._kwargs = kwargs
+
+    def __get__(self, obj, cls):
+        # Return None if our object is immutable and the data is not present.
+        if not obj.mutable and self._adt not in obj.data_dict:
+            return None
+        print(self._adt, self._cls, repr(obj))
+        bound_class = self._cls(obj, advertising_data_type=self._adt, **self._kwargs)
+        setattr(obj, self._attribute_name, bound_class)
+        obj.data_dict[self._adt] = bound_class
+        return bound_class
+
+    # TODO: Add __set_name__ support to CircuitPython so that we automatically tell the descriptor
+    # instance the attribute name it has and the class it is on.
+
+class Advertisement:
```

Comment:
They can access data_dict if they like. They can also `bytes()` it to get the raw data back.

I didn't want the top-level object to handle all fields because some of them require sub-parsing to make any sense. That parsing happens in advertising type specific modules that can be scanned for explicitly. See `examples/ble_detailed_scan.py` for an example that does this for services.